server.ts

```diff
@@ -0,0 +1,318 @@
+}
+
+// Displays storage contents.
+function displayStorage() {
+  console.log('new storage state:');
+  storage.forEach((channel: Channel, channelId: string) => {
+    console.log('\t' + channelId + ': (' + channel.expectedPresenceMembers + ')');
+    channel.members.forEach((member: Member, _) => {
+      console.log('\t\t' + member.clientId + '-' + member.connectionId + ': ' + member.latestTimestamp);
+    })
+  });
+}
+
+// Sets a timer to check presence count for a channel. Resets any active counter.
+function setupMemberCountCheckTimeout(channel: Channel, channelId: string) {
+  if (channel.memberCountCheckTimeout) {
+    clearTimeout(channel.memberCountCheckTimeout);
+  }
+  channel.memberCountCheckTimeout = setTimeout(() => checkPresenceMemberCount(channelId), 5000);
+}
+
+// Process an occupancy update.
+function occupancyUpdate(channelId: string, occupancy: Occupancy) {
+  if (!storage.has(channelId)) {
+    return;
+  }
+  const channel = storage.get(channelId)!;
+  channel.expectedPresenceMembers = occupancy.metrics.presenceMembers;
+  setupMemberCountCheckTimeout(channel, channelId); // Check presence member count in a few seconds.
+  console.log('> occupancy update: ' + channelId + '=' + occupancy.metrics.presenceMembers);
+  displayStorage();
+}
+
+// Process a presence update.
+// updateMemberCount is set to true when we process a presence set returned by the REST request.
+// In that case we are not expecting an occupancy update, so we can directly update the member count.
+function presenceUpdate(channelId: string, update: Ably.Types.PresenceMessage, updateMemberCount: boolean) {
+  console.log('> presence update: ' + channelId);
+  // Ignore updates with a timestamp older than ours.
+  const memberKey = new MemberKey(update.clientId, update.connectionId);
+  if (storage.has(channelId)) {
+    const channel = storage.get(channelId)!;
+    if (channel.members.has(memberKey.hash)) {
+      const member = channel.members.get(memberKey.hash)!;
+      if (update.timestamp < member.latestTimestamp) {
+        return;
+      }
+    }
+  }
+  switch (update.action) {
+    case "present":
+    case "enter":
+    case "update": {
+      const channel = emplace(storage, channelId, () => new Channel());
+      const member = emplace(channel.members, memberKey.hash, () => new Member(update.clientId, update.connectionId));
+      member.latestTimestamp = update.timestamp;
+      setupMemberCountCheckTimeout(channel, channelId); // Check presence member count in a few seconds.
+      if (updateMemberCount) {
+        channel.expectedPresenceMembers = channel.members.size;
+      }
+      displayStorage();
+      break;
+    }
+    case "leave": {
+      // Do some cleanup if needed (remove empty channels).
+      if (!storage.has(channelId)) {
+        break;
+      }
+      const channel = storage.get(channelId)!;
+      if (!channel.members.has(memberKey.hash)) {
+        break;
+      }
+      channel.members.delete(memberKey.hash);
```

Comment:
don't we need to tombstone the delete so that if an earlier enter, or update comes through then we don't subsequently add it back again?

Reply:
That should not be possible since we check the update's timestamp, so any earlier update coming after will be ignored.

Reply:
Also I believe this tombstone would prevent the same user from leaving and then entering the same channel with the same connection ID.